keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,1128 @@
+implementation choice means that:
+* The control plane (kube-controller-manager in this case) is only concerned
+  with the field `routingPreference` and populates EndpointSlice hints based
+  on its value 
+* The data plane (kube-proxy in this case) is only concerned with the Hints
+  populated in the EndpointSlice and the fields `internal/externalTrafficPolicy`
+  to make routing decisions.
+* Neither the control plane nor the data plane looks at the others field.
+
+NOTE: The fact that EndpointSlice hints are not expected to be implemented by
+  all data planes is not of concern here, because `routingPreference` is anyways
+  supposed to be implementation dependent. Since we know that the kube-proxy
+  data plane will respect the Hints, it’s sufficient for the kube-proxy and
+  kube-controller-manager based implementations to say that we do support the
+  standard heuristics.
+
+### Status Reporting
+
+To provide clear status updates about the routing preferences to the user, the
+EndpointSlice controller (which is acting as the control plane) will update the
+Service status with the following conditions (inspired by Gateway API conditions)
+
+* RoutingPreferenceAccepted
+  * **Type:** `RoutingPreferenceAccepted` 
+  * **Description:** 
+    * Indicates whether a control plane component recognized and successfully
+      parsed the `routingPreference` value. A `False` status typically suggests
+      a configuration issue (e.g., an unsupported or malformed preference).
+    * The EndpointSlice controller will set this status to `True` if it
+      recognizes the `routingPreference` value as one it explicitly supports. 
+
+* RoutingPreferenceProgrammed
+  * **Type:**  `RoutingPreferenceProgrammed`
+  * **Description:** 
+    * This condition indicates whether they `routingPreference` has generated
+      some configuration that is assumed to be ready soon in the underlying data
+      plane.
+    * The EndpointSlice controller will set this status to `True` if it
+      successfully populated the EndpointSlice hints based on the
+      `routingPreference`.
+
+Note that the EndpointSlice controller and kube-proxy implementation does not
+currently provide a condition denoting acknowledgment from the dataplane. In the
+future when this is possible, another condition like the following could be
+used:
+
+* RoutingPreferenceReady
+  * **Type:** `RoutingPreferenceReady`
+  * **Description:** Confirms that the dataplane has received the hints,
+    acknowledged them, and configured itself accordingly.
+
+#### Condition usage by other implementations
+
+Other implementations supporting `routingPreference` **should** adopt a domain
+prefixing strategy for their condition types. This means prefixing condition
+types with a domain string (e.g., `my.domain.io/RoutingPreferenceAccepted`) to
+prevent conflicts when multiple control planes (like the default EndpointSlice
+controller) are present.
+
+### Choice of field name
```

Comment:
Naming debate thread:

I don't love it.  "routing" feels really low-level.  We know this is about locality in practice, but maybe that is too concrete?  I guess, in theory, someone could use different criteria.

readers: ideas welcome

Reply:
How do folks feel about `trafficDistribution` with `Prefer` keyword in the field values? (Assuming `trafficDistributionPreference` might be too long)

Reply:
agree that "routing" implies the wrong thing

> with `Prefer` keyword in the field values

moving "prefer" from the field name to the values opens up the possibility of having `Require` (or at least "stronger-than-Prefer") values in the future

Reply:
Final names can be argued in impl, but we will need to come back and update this.

Reply:
My tip: come up with really clear name for the feature gate, one that communicates what the feature is and does. Then see if that helps us name the API fields.